rasa/architecture_prototype/graph_components.py

```diff
@@ -106,63 +115,377 @@ def generate(
+                        self._num_collisions_resolved += 1
+                    else:
+                        self._num_collisions_ignored += 1
+
+                else:
+                    raise ValueError(
+                        f"Expected added message to be consistent. {key} already maps to "
+                        f"{existing_message}, but we want to add {message} now."
+                    )
+            else:
+                self._num_collisions_ignored += 1
+        else:
+            self._table[key] = message
+
+    def add_all(self, messages: Iterable[Message]) -> None:
+        """Adds the given message to the lookup table.
+
+        Args:
+          messages: the messages that we want to add to the lookup table
+        Raises:
+          ValueError if we cannot create a key for one of the given messages or if
+          adding a message creates a collision that we cannot resolve
+        """
+        for message in messages:
+            self.add(message)
+
+    def lookup_features(
+        self, sub_state: SubState, attributes: Optional[List[Text]] = None
+    ) -> Dict[Text, List[Features]]:
+        """Looks up all features for the given substate.
+
+        Note that here we do **not** impose any restrictions on the attributes that
+        may be included in the given sub_state.
+        There might be be multiple messages in the lookup table that contain features
+        relevant for the given substate, e.g. this is the case if `TEXT` and
+        `INTENT` are present in the given message.
+
+        Args:
+          sub_state: substate for which we want to lookup all relevant Messages
+          attributes: if not None, this specifies the list of the attributes of the
+            `Features` that we're interested in (i.e. all other `Features` contained
+            in the relevant messages will be ignored)
+        Returns:
+          a dictionary that maps all the (requested) attributes for which we found
+          `Features` to a list of these `Features`
+        Raises:
+          - a `ValueError` if any of the lookup table keys that can be generated from
+            the given substate is not contained in the lookup table
+          - a `RuntimeError` if an inconsistency in the lookup table is detected;
+            more precisely, if features for the same attribute are found in two
+            looked up messages
+        """
+        features = {}
+        # get all keys whose values in the lookup table contain features for the
+        # given substate
+        key_attributes = set(sub_state.keys()).intersection(self.KEY_ATTRIBUTES)
+        for key_attribute in key_attributes:
+            key = self._build_key(sub_state={key_attribute: sub_state[key_attribute]})
+            message = self._table.get(key)
+            if not message:
+                raise ValueError(
+                    f"Unknown key {key}. Cannot retrieve features for substate "
+                    f"{sub_state}"
+                )
```

Comment:
It is also used during prediction. This is a "this should never happen - if we use the lookup table correctly" kind of error message. E.g. if `applied_events` started to generate new events that are no in the event list that I generate the lookup table from, then this would be raised.